pkg/kubectl/cmd/patch.go

```diff
@@ -200,14 +200,17 @@ func RunPatch(f cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []strin
 			}
 			count++
 
-			if err := info.Refresh(patchedObj, true); err != nil {
+			//info.Object is updated by the info.Refresh call.
+			//we should compute old data prior to refresh
+			oldData, err := json.Marshal(info.Object)
```

Comment:
I have tried to move it here. But it doesn't work for all of the cases.
You should make sure all of these work as expected:
```
kubectl patch -f deployment.yaml -p '{}'
not patched
kubectl patch deployment nginx -p '{}'
not patched
kubectl patch -f deployment.yaml -p '{"spec":{"replicas":1}}'
patched
kubectl patch -f deployment.yaml -p '{"spec":{"replicas":2}}'
patched
```